File: nodejs/snapshot.js
const angle = require("./math/angle");
const Point = require("./math/point");

class Snapshot {
    data;
    color;
    robot;
    ball;

    constructor(data) {
        this.data = data;
        this.color = data.robot.color;
        this.processRobotSensors(data.robot);
        this.processBallSignal(data.ball);
        this.mergeTeamData(data.team);
    }

    isBallDetected() {
        return this.data.ball != null;
    }

    processRobotSensors(robot_data) {
        let x, y, cx, cy;
        [x, y] = robot_data.gps;
        [cx, cy] = robot_data.compass;
        this.robot = {
            name: robot_data.name,
            index: robot_data.index,
            position: new Point(x, y),
            rotation: angle.radians(Math.atan2(cx, cy) + Math.PI/2)
        };
    }

    processBallSignal(ball_data) {
        if (!ball_data) return null;
        let dist = Math.sqrt(1/ball_data.strength);
        let x, y;
        [x, y] = ball_data.direction;
        let da = angle.radians(Math.atan2(y, x));
        let a = angle.radians(this.robot.rotation + da);
        let dx = Math.sin(a) * dist;
        let dy = Math.cos(a) * -1 * dist;
        let bx = this.robot.position.x + dx;
        let by = this.robot.position.y + dy;
        this.ball = {
            position: new Point(bx, by)
        }
    }

    mergeTeamData(team_data) {
        if (!this.ball) {
            if (team_data && team_data.length > 0) {
                this.ball = team_data[0];
            }
        }
    }
}

module.exports = Snapshot;

File: python/snapshot.py
from dataclasses import dataclass
from math_utils.angle import radians
from math_utils.point import Point
import math

@dataclass
class RobotData:
    name: str
    index: int
    position: Point
    rotation: float

@dataclass
class BallData:
    position: Point


class Snapshot:
    data = None
    color = None
    robot = None
    ball = None

    def __init__(self, data):
        self.data = data
        self.color = data["robot"]["color"]
        self.processRobotSensors(data["robot"])
        self.processBallSignal(data.get("ball"))
        self.mergeTeamData(data.get("team"))

    def isBallDetected(self):
        return self.data.get("ball") != None

    def processRobotSensors(self, robot_data):
        x, y, _ = robot_data["gps"]
        cx, cy, _ = robot_data["compass"]
        self.robot = RobotData(
            name=robot_data["name"],
            index=robot_data["index"],
            position=Point(x, y),
            rotation=radians(math.atan2(cx, cy) + math.pi/2))

    def processBallSignal(self, ball_data):
        if ball_data == None: return None
        dist = math.sqrt(1/ball_data["strength"])
        x, y, _ = ball_data["direction"]
        da = radians(math.atan2(y, x))
        a = radians(self.robot.rotation + da)
        dx = math.sin(a) * dist
        dy = math.cos(a) * -1 * dist
        bx = self.robot.position.x + dx
        by = self.robot.position.y + dy
        self.ball = BallData(position=Point(bx, by))

    def mergeTeamData(self, team_data):
        if self.ball == None:
            if team_data and len(team_data) > 0:
                x, y = team_data[0]
                self.ball = BallData(position=Point(x, y))

